### PR TITLE
[WIP] feat: add limited resource for critical pod in all namespaces - part1

### PR DESCRIPTION
**Pull request by draveness**
/kind feature
/priority important-soon
/sig scheduling

**What this PR does / why we need it**:

+ Remove limitation of critical pod creation in kube-system namespace in pod priority admission plugin
+  Create a AdmissionConfiguration object with limitedResources to prevent creation of system critical pods in all namespaces
+  Add a default quota with scope selector to allow critical pods to be created in kube-system namespace only

**Which issue(s) this PR fixes**:

ref: https://github.com/kubernetes/kubernetes/issues/82480

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/blob/master/keps/sig-scheduling/20190409-resource-quota-ga.md
```

### CODE REVIEW

**Comment by draveness** [comment]
The integration tests keep failing https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/82575/pull-kubernetes-integration/1174599081327071233/:

```go
--- FAIL: TestGracefulShutdown (11.36s)
    testserver.go:143: runtime-config=map[api/all:true]
    testserver.go:144: Starting kube-apiserver on port 36479...
    testserver.go:160: Waiting for /healthz to be ok...
    graceful_shutdown_test.go:51: server should be blocking request for data in body
    graceful_shutdown_test.go:66: server should answer
    graceful_shutdown_test.go:73: shutting down server
    graceful_shutdown_test.go:76: server should fail new requests
    graceful_shutdown_test.go:88: server should answer pending request
    graceful_shutdown_test.go:95: respErr {<nil> {441 0 }}
    graceful_shutdown_test.go:96: respErr http2: server sent GOAWAY and closed the connection; LastStreamID=441, ErrCode=NO_ERROR, debug=""
```

File: test/integration/master/graceful_shutdown_test.go [hunk 1]
```diff
@@ -92,6 +92,8 @@ func TestGracefulShutdown(t *testing.T) {
 	}
 	body.Close()
 	respErr := <-respErrCh
+	t.Logf("respErr %v", respErr)
+	t.Logf("respErr %v", respErr.err)
 	if err != nil {
```
Comment (damemi) [inline test/integration/master/graceful_shutdown_test.go]: This should be `if respErr.err != nil`

Comment (damemi) [reply test/integration/master/graceful_shutdown_test.go]: Looking at the failures, it seems to be a nil pointer somewhere around here:
```
--- FAIL: TestGracefulShutdown (12.71s)
    testserver.go:143: runtime-config=map[api/all:true]
    testserver.go:144: Starting kube-apiserver on port 43375...
    testserver.go:160: Waiting for /healthz to be ok...
    graceful_shutdown_test.go:51: server should be blocking request for data in body
    graceful_shutdown_test.go:66: server should answer
    graceful_shutdown_test.go:73: shutting down server
    graceful_shutdown_test.go:76: server should fail new requests
    graceful_shutdown_test.go:88: server should answer pending request
    graceful_shutdown_test.go:95: respErr {<nil> 0xc0492fe4b0}
    graceful_shutdown_test.go:96: respErr read tcp 127.0.0.1:43316->127.0.0.1:43375: read: connection reset by peer
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x40 pc=0x386d408]

goroutine 50811 [running]:
testing.tRunner.func1(0xc083d65600)
	/usr/local/go/src/testing/testing.go:830 +0x392
panic(0x3ee3360, 0xaef2450)
	/usr/local/go/src/runtime/panic.go:522 +0x1b5
k8s.io/kubernetes/test/integration/master.TestGracefulShutdown(0xc083d65600)
	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/integration/master/graceful_shutdown_test.go:100 +0xb08
testing.tRunner(0xc083d65600, 0x48901b0)
	/usr/local/go/src/testing/testing.go:865 +0xc0
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:916 +0x35a
FAIL	k8s.io/kubernetes/test/integration/master	194.552s
```

Comment (draveness) [reply test/integration/master/graceful_shutdown_test.go]: I'll try this approach, thanks Mike

File: test/integration/master/graceful_shutdown_test.go [hunk 2]
```diff
@@ -92,7 +92,8 @@ func TestGracefulShutdown(t *testing.T) {
 	}
 	body.Close()
 	respErr := <-respErrCh
-	if err != nil {
+	t.Logf("respErr %v", respErr)
+	if respErr.err != nil {
 		t.Fatal(err)
```
Comment (damemi) [inline test/integration/master/graceful_shutdown_test.go]: This should also be `t.Fatal(respErr.err)`, but I'm not sure why it's returning an error or what that error is

Comment (draveness) [reply test/integration/master/graceful_shutdown_test.go]: > This should also be `t.Fatal(respErr.err)`, but I'm not sure why it's returning an error or what that error is

It doesn't seem to affect the result.

Comment (draveness) [reply test/integration/master/graceful_shutdown_test.go]: It's quite confusing why we need to change the test...

Comment (damemi) [reply test/integration/master/graceful_shutdown_test.go]: These are mistakes in the test that we are only noticing now because the test is failing, so we need to fix it to check the proper error and then report the proper error on a failure. Right now, the test is failing, but the `fatal` call is only reporting `<nil>` as the error, so fixing the check and reporting is important

**Comment by ahg-g** [comment]
Hi @draveness, any progress here, do you need help? 

/milestone 1.17

**Comment by draveness** [comment]
Per offline discussion with @ravisantoshgudimetla, I'm going to close this by. superseded by #76310 

/close

File: plugin/pkg/admission/priority/admission_test.go [hunk 3]
```diff
@@ -626,7 +626,7 @@ func TestPodAdmission(t *testing.T) {
 			[]*scheduling.PriorityClass{systemClusterCritical},
 			*pods[7],
 			scheduling.SystemCriticalPriority,
-			true,
+			false,
```
Comment (sttts) [inline plugin/pkg/admission/priority/admission_test.go]: why this?

File: plugin/pkg/admission/priority/admission.go [hunk 4]
```diff
@@ -177,19 +180,27 @@ func (p *Plugin) admitPod(a admission.Attributes) error {
 	}
 
 	if operation == admission.Create {
-		var priority int32
-		var preemptionPolicy *apiv1.PreemptionPolicy
+		var (
+			priority         int32
+			preemptionPolicy *apiv1.PreemptionPolicy
+		)
```
Comment (sttts) [inline plugin/pkg/admission/priority/admission.go]: ugly

File: plugin/pkg/admission/priority/admission.go [hunk 5]
```diff
@@ -177,19 +180,27 @@ func (p *Plugin) admitPod(a admission.Attributes) error {
...
 		if len(pod.Spec.PriorityClassName) == 0 {
-			var err error
-			var pcName string
+			var (
+				err    error
+				pcName string
+			)
```
Comment (sttts) [inline plugin/pkg/admission/priority/admission.go]: ugly

File: plugin/pkg/admission/resourcequota/config.go [hunk 6]
```diff
@@ -21,11 +21,11 @@ import (
 	"io"
 	"io/ioutil"
 
+	v1 "k8s.io/api/core/v1"
```
Comment (sttts) [inline plugin/pkg/admission/resourcequota/config.go]: corev1

File: test/integration/master/graceful_shutdown_test.go [hunk 7]
```diff
@@ -92,8 +92,9 @@ func TestGracefulShutdown(t *testing.T) {
 	}
 	body.Close()
 	respErr := <-respErrCh
-	if err != nil {
-		t.Fatal(err)
+	t.Logf("respErr %v", respErr)
+	if respErr.err != nil {
+		t.Fatal(respErr.err)
```
Comment (sttts) [inline test/integration/master/graceful_shutdown_test.go]: can you move this into a separate PR? It was obviously wrong.
